### PR TITLE
Fix for #48, allow limiting dates in the calendar

### PR DESCRIPTION
**Pull request by aryaxt**
Thanks for the awesome component.
This PR allows setting a start or end range, 
I'm also working on support for date range selection, will open another PR soon

### CODE REVIEW

File: RSDayFlow/RSDFDatePickerDayCell.m [hunk 1]
```diff
@@ -213,49 +213,58 @@ - (void)updateSubviews
...
+					self.dateLabel.textColor = [self selectedDayLabelTextColor];
+					self.selectedDayImageView.image = [self selectedDayImage];
+				} else {
+					self.dateLabel.font = [self selectedTodayLabelFont];
+					self.dateLabel.textColor = [self selectedTodayLabelTextColor];
+					self.selectedDayImageView.image = [self selectedTodayImage];
+				}
+			}
+			
+			if (self.marked) {
```
Comment (aryaxt) [inline RSDayFlow/RSDFDatePickerDayCell.m]: should this be outside of the main if/else?

**Comment by ruslanskorb** [comment]
Hi @aryaxt,
Thanks for the PR!
I had a quick look at it and did not see the changes for the paging mode.
I'll look at it more deeply as soon as possible.

**Comment by aryaxt** [comment]
@ruslanskorb yes, there are two date picker views in the sample, I only set the date limit for one of them in the demo, the other should still have an infinite range

**Comment by aryaxt** [comment]
I just pushed a change with a small issue where disabled text color had priority over notThisMonth text color. Will integrate this into my project and do lots of testing tomorrow

**Comment by aryaxt** [comment]
@ruslanskorb My testing is done, it's all good :)

**Comment by ruslanskorb** [comment]
I have a lot of comments, but I don't have time to write them today. I'll do it tomorrow.

**Comment by ruslanskorb** [comment]
[1] If we set the start and end dates in the paging mode, and scroll to the top month or the bottom month, the application will crash. To fix it we need to check for the start and end dates in the method `scrollViewWillBeginDragging:`. (RSDFDatePickerView.m)

```obj-c
- (void)scrollViewWillBeginDragging:(UIScrollView *)scrollView
{
    dispatch_async(dispatch_get_main_queue(), ^{
        if (self.isPagingEnabled) {
            if (!self.startDate && scrollView.contentOffset.y < CGRectGetHeight(scrollView.bounds) * 2) {
                [self appendPastDates];
            }
            
            if (!self.endDate && scrollView.contentOffset.y > scrollView.contentSize.height - CGRectGetHeight(scrollView.bounds) * 2) {
                [self appendFutureDates];
            }
        }
    });
}
```

[2] In addition we need to check for the start and end dates in the method `scrollViewWillEndDragging:withVelocity:targetContentOffset:` to exclude non-existent months. (RSDFDatePickerView.m)

[3] And we need to update `targetContentOffset` for the last month in case when it has height less than the height of the date picker view. (RSDFDatePickerView.m)

```obj-c
- (void)scrollViewWillEndDragging:(UIScrollView *)scrollView withVelocity:(CGPoint)velocity targetContentOffset:(inout CGPoint *)targetContentOffset
{
    if (self.isPagingEnabled) {
        NSArray *sortedIndexPathsForVisibleItems = [[self.collectionView indexPathsForVisibleItems] sortedArrayUsingComparator:^NSComparisonResult(NSIndexPath *obj1, NSIndexPath * obj2) {
            return obj1.section > obj2.section;
        }];
        
        NSUInteger visibleSection;
        NSUInteger nextSection;
        if (velocity.y > 0.0) {
            visibleSection = [[sortedIndexPathsForVisibleItems firstObject] section];
            nextSection = visibleSection + 1;
        } else if (velocity.y < 0.0) {
            visibleSection = [[sortedIndexPathsForVisibleItems lastObject] section];
            nextSection = visibleSection - 1;
        } else {
            visibleSection = [sortedIndexPathsForVisibleItems[sortedIndexPathsForVisibleItems.count / 2] section];
            nextSection = visibleSection;
        }
        
        if (self.startDate) {
            if (nextSection < [self sectionForDate:self.startDate]) {
                return;
            }
        }
        if (self.endDate) {
            if (nextSection > [self sectionForDate:self.endDate]) {
                return;
            }
        }
        
        CGRect headerRect = [self frameForHeaderForSection:nextSection];
        CGPoint topOfHeader = CGPointMake(0, headerRect.origin.y - self.collectionView.contentInset.top);
        CGFloat maxYContentOffset = self.collectionView.contentSize.height - CGRectGetHeight(self.collectionView.bounds);
        if (topOfHeader.y > maxYContentOffset) {
            topOfHeader.y = maxYContentOffset;
        }
        
        *targetContentOffset = topOfHeader;
        
        scrollView.decelerationRate = UIScrollViewDecelerationRateFast;
    }
}
```

[4] We should cancel scroll to the date that is out of range. (RSDFDatePickerView.m)

```obj-c
- (void)scrollToDate:(NSDate *)date animated:(BOOL)animated
{
    if (self.startDate && [date compare:self.startDate] == NSOrderedAscending) {
        return;
    }
    if (self.endDate && [date compare:self.endDate] == NSOrderedDescending) {
        return;
    }
    ...
}
```

[5] We should use non-deprecated calendar units. (RSDFDatePickerView.m)

```obj-c
- (NSDate *)dateWithFirstDayOfMonth:(NSDate *)date
{
	NSDateComponents *dateComponents = [self.calendar components:(NSCalendarUnitYear | NSCalendarUnitMonth | NSCalendarUnitDay) fromDate:date];
	dateComponents.day = 1;
	return [self.calendar dateFromComponents:dateComponents];
}

- (NSDate *)dateByMovingToEndOfMonth:(NSDate *)date
{
	NSDateComponents *components = [self.calendar components:(NSCalendarUnitYear | NSCalendarUnitMonth | NSCalendarUnitDay) fromDate:date];
	
	components.month = components.month + 1;
	return [self.calendar dateFromComponents:components];
}
```

[6] We don't need to check for the start and end dates in the method `shiftDatesByComponents:`. It is already done in the methods `pickerCollectionViewWillLayoutSubviews:` and `scrollViewWillBeginDragging:`. (RSDFDatePickerView.m)

```obj-c
_fromDate = [self pickerDateFromDate:[self.calendar dateByAddingComponents:components toDate:[self dateFromPickerDate:self.fromDate] options:0]];
_toDate = [self pickerDateFromDate:[self.calendar dateByAddingComponents:components toDate:[self dateFromPickerDate:self.toDate] options:0]];
```

[7] I would like to rename the property `disabled` to `outOfRange` and the method `disabledDayLabelTextColor` to `outOfRangeDayLabelTextColor `. (RSDFDatePickerDayCell.h+m)

```obj-c
/**
 A Boolean value that determines whether the cell is out of the start/end date range.
 */
@property (nonatomic, getter = isOutOfRange) BOOL outOfRange;

/**
 The text color for the label of the day that is out of the start/end date range. Default value is [UIColor colorWithRed:184/255.0f green:184/255.0f blue:184/255.0f alpha:1.0f].
 
 @discussion Can be overridden in subclasses for customization.
 */
- (UIColor *)outOfRangeDayLabelTextColor;
```

[8] We should hide `selectedDayImageView`, `overlayImageView` and `markImageView` for dates that are out of range. (RSDFDatePickerDayCell.m)

```obj-c
- (void)updateSubviews
{
    self.selectedDayImageView.hidden = !self.isSelected || self.isNotThisMonth || self.isOutOfRange;
    self.overlayImageView.hidden = !self.isHighlighted || self.isNotThisMonth || self.isOutOfRange;
    self.markImageView.hidden = !self.isMarked || self.isNotThisMonth || self.isOutOfRange;
    ...
}
```

[9] We should update the font for the dates that are out of range. (RSDFDatePickerDayCell.m)

```obj-c
if (self.isOutOfRange) {
    self.dateLabel.font = [self dayLabelFont];
    self.dateLabel.textColor = [self outOfRangeDayLabelTextColor];
}
```

[10] By default, we assume that the user is using dates without the time. This assumption allows us to avoid unnecessary conversions and thereby increase performance. So, we can remove the method `dateWithoutTimeComponents:`. (RSDFDatePickerView.m)

[11] We don't need to use the method `isCellEnabledAtIndexPath:indexPath:`. Instead we can use the following code in the method `collectionView:cellForItemAtIndexPath:`: (RSDFDatePickerView.m)

```obj-c
if ((self.startDate && [cellDate compare:self.startDate] == NSOrderedAscending) ||
    (self.endDate && [cellDate compare:self.endDate] == NSOrderedDescending)) {
    cell.outOfRange = YES;
} else {
    cell.outOfRange = NO;
}
```

And then we can add simple check in the method `collectionView:shouldHighlightItemAtIndexPath:` and the method `collectionView:shouldSelectItemAtIndexPath:`: (RSDFDatePickerView.m)

```obj-c
RSDFDatePickerDayCell *cell = ((RSDFDatePickerDayCell *)[collectionView cellForItemAtIndexPath:indexPath]);
if (cell.isNotThisMonth || cell.isOutOfRange) {
    return NO;
}
```

It will also allow us not to change the value of the property `userInteractionEnabled`. Just remove it from the code. (RSDFDatePickerDayCell.m)

[12] Please use the RSDayFlow style for formatting your code:
* Other braces (`if`/`else`/`switch`/`while` etc.) always open on the same line as the statement but close on a new line.

**Preferred:**
```objc
if (user.isHappy) {
    // Do something
} else {
    // Do something else
}
```

**Not Preferred:**
```objc
if (user.isHappy)
{
    // Do something
}
else {
    // Do something else
}
```

**Comment by aryaxt** [comment]
@ruslanskorb I'm busy during the week, will go over it this weekend, should I combine the changes for range-picking in this PR? or do you want a separate PR for that? https://github.com/ruslanskorb/RSDayFlow/compare/master...aryaxt:AG/SupportForPickingADateRange

**Comment by ruslanskorb** [comment]
@aryaxt Okay, no probs!
I would like a separate PR for range-picking.
Thanks!

**Comment by aryaxt** [comment]
@ruslanskorb 

[9] I think it would be nice to also allow changing the font for an out of range date. For instance, in my project a regular date has a normal font, and an out of range date has a light font.

[10] Maybe we can remove the time component when the start/end dates are set through the init method? That would ensure that doesn't matter what the time is, we'll enable the correct range.

**Comment by ruslanskorb** [comment]
@aryaxt 

[3] Good catch!

[6] As I pointed out above it is already done in the done in the methods `pickerCollectionViewWillLayoutSubviews:` and `scrollViewWillBeginDragging:`. That's enough, you can be sure.

[9] Agree. We can add the font for an out of range date.

[10]  Agree. Good suggestion!

**Comment by aryaxt** [comment]
@ruslanskorb Those methods don't stop pickerCollectionViewWillLayoutSubviews to get called if one end is open. It'll crash without that code. Checkout the code, set the startDate to current date and leave the end date open, it'll crash.
```
[[RSDFDatePickerView alloc] initWithFrame:self.view.bounds calendar:self.calendar startDate:[NSDate date] endDate:nil];
```

File: RSDayFlow/RSDFDatePickerView.m [hunk 2]
```diff
@@ -700,7 +770,9 @@ - (UICollectionReusableView *)collectionView:(UICollectionView *)collectionView
 
 - (BOOL)collectionView:(UICollectionView *)collectionView shouldHighlightItemAtIndexPath:(NSIndexPath *)indexPath
 {
-    if (((RSDFDatePickerDayCell *)[collectionView cellForItemAtIndexPath:indexPath]).isNotThisMonth) {
+    RSDFDatePickerDayCell *cell = (RSDFDatePickerDayCell *)[collectionView cellForItemAtIndexPath:indexPath];
```
Comment (ruslanskorb) [inline RSDayFlow/RSDFDatePickerView.m]: Please remove doublet after `if ([self.delegate respondsToSelector:@selector(datePickerView:shouldHighlightDate:)]) {`

File: RSDayFlow/RSDFDatePickerView.m [hunk 3]
```diff
@@ -727,12 +799,13 @@ - (void)collectionView:(UICollectionView *)collectionView didUnhighlightItemAtIn
 
 - (BOOL)collectionView:(UICollectionView *)collectionView shouldSelectItemAtIndexPath:(NSIndexPath *)indexPath
 {
-    if (((RSDFDatePickerDayCell *)[collectionView cellForItemAtIndexPath:indexPath]).isNotThisMonth) {
+	RSDFDatePickerDayCell *cell = ((RSDFDatePickerDayCell *)[self.collectionView cellForItemAtIndexPath:indexPath]);
```
Comment (ruslanskorb) [inline RSDayFlow/RSDFDatePickerView.m]: Spaces instead of tab.

File: RSDayFlow/RSDFDatePickerView.m (hunk 2 above)
Comment (ruslanskorb) [inline RSDayFlow/RSDFDatePickerView.m]: Please remove doublet after `if ([self.delegate respondsToSelector:@selector(datePickerView:shouldHighlightDate:)]) {`

**Comment by ruslanskorb** [comment]
@aryaxt The PR will be merged after these small fixes. Thanks!

**Comment by aryaxt** [comment]
@ruslanskorb done

**Comment by ruslanskorb** [comment]
@aryaxt Merged!

**Comment by aryaxt** [comment]
awesome

**Comment by ruslanskorb** [comment]
@aryaxt Will you open another PR for date range selection?

**Comment by aryaxt** [comment]
@ruslanskorb  Yes, just doing some testing and review before opening the PR

**Comment by ruslanskorb** [comment]
@aryaxt OK

**Pull request merged**